tools/make_patch.c

```diff
@@ -0,0 +1,476 @@
+#define PROGRAM_NAME "make_patch"
+#define USAGE_OPTS "labels.sym constants.out patched.gbc original.gbc vc.patch.template vc.patch"
+
+#include "common.h"
+
+#include <ctype.h>
+
+struct Buffer {
+	size_t item_size;
+	size_t size;
+	size_t capacity;
+	char *data;
+};
+
+struct Patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct Symbol {
+	struct Symbol *next;
+	unsigned int address;
+	unsigned int offset;
+	char name[]; // VLA
+};
+
+struct Buffer *create_buffer(size_t item_size) {
+	struct Buffer *buffer = xmalloc(sizeof(*buffer));
+	buffer->item_size = item_size;
+	buffer->size = 0;
+	buffer->capacity = 0x10;
+	buffer->data = xmalloc(buffer->capacity * item_size);
+	return buffer;
+}
+
+void append_to_buffer(struct Buffer *buffer, const void *item) {
+	if (buffer->size >= buffer->capacity) {
+		buffer->capacity = (buffer->capacity + 1) * 2;
+		buffer->data = xrealloc(buffer->data, buffer->capacity * buffer->item_size);
+	}
+	memcpy(buffer->data + (buffer->size++ * buffer->item_size), item, buffer->item_size);
+}
+
+void free_buffer(struct Buffer *buffer) {
+	free(buffer->data);
+	free(buffer);
+}
+
+int get_address_type_limit(int address) {
+	return address >= 0x8000 && address < 0xa000 ? 0xa000 // VRAM
+	     : address >= 0xa000 && address < 0xc000 ? 0xc000 // SRAM
+	     : address >= 0xc000 && address < 0xf000 ? 0xd000 // WRAM
+	     : address >= 0xff80 && address < 0xffff ? 0x10000 // HRAM
+	     : 0x4000; // ROM
```

Comment:
Since that's wrong for e.g. OAM, why not simply use "cap" checks?
```c
int get_address_type_limit(int address) {
    return address < 0x4000 ? 0x4000 // ROM0
         : address < 0x8000 ? 0x8000 // ROMX
         : address < 0xa000 ? 0xa000 // VRAM
         : address < 0xc000 ? 0xc000 // SRAM
         : address < 0xe000 ? 0xe000 // WRAM
         : 0xffff; // Assuming HRAM
}
```